aleph-client/src/pallets/multisig.rs

```diff
@@ -71,3 +129,388 @@ impl<S: SignedConnectionApi> MultisigUserApi for S {
+    threshold: MultisigThreshold,
+}
+
+impl MultisigParty {
+    /// Create new party from `signatories` and `threshold`.
+    ///
+    /// `signatories` can contain duplicates and doesn't have to be sorted. However, there must be
+    /// at least 2 unique accounts. There is also a virtual upper bound - `MaxSignatories` constant.
+    /// It isn't checked here (since it requires client), however, using too big party will fail
+    /// when performing any chain interaction.
+    ///
+    /// `threshold` must be between 2 and number of unique accounts in `signatories`. For threshold
+    /// 1, use special method `MultisigUserApi::as_multi_threshold_1`.
+    pub fn new(signatories: &[AccountId], threshold: MultisigThreshold) -> AnyResult<Self> {
+        let mut sorted_signatories = signatories.to_vec();
+        sorted_signatories.sort();
+        sorted_signatories.dedup();
+
+        ensure!(
+            sorted_signatories.len() > 1,
+            "There must be at least 2 different signatories"
+        );
+        ensure!(
+            sorted_signatories.len() >= threshold as usize,
+            "Threshold must not be greater than the number of unique signatories"
+        );
+        ensure!(
+            threshold >= 2,
+            "Threshold must be at least 2 - for threshold 1, use `as_multi_threshold_1`"
+        );
+
+        Ok(Self {
+            signatories: sorted_signatories,
+            threshold,
+        })
+    }
+
+    /// The multisig account derived from signatories and threshold.
+    ///
+    /// This method is copied from the pallet, because:
+    ///  -  we don't want to add a new dependency
+    ///  -  we cannot instantiate pallet object here anyway (the corresponding functionality exists
+    ///     as pallet's method rather than standalone function)
+    pub fn account(&self) -> AccountId {
+        let entropy =
+            (b"modlpy/utilisuba", &self.signatories, &self.threshold).using_encoded(blake2_256);
+        Decode::decode(&mut TrailingZeroInput::new(entropy.as_ref()))
+            .expect("infinite length input; no invalid inputs for type; qed")
+    }
+}
+
+/// A context in which ongoing signature aggregation is performed.
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub struct Context {
+    /// The entity for which aggregation is being made.
+    party: MultisigParty,
+    /// Derived multisig account (the source of the target call).
+    author: AccountId,
+
+    /// Pallet's coordinate for this aggregation.
+    timepoint: Timepoint,
+    /// Weight limit when dispatching the call.
+    max_weight: Weight,
+
+    /// The target dispatchable, if already provided.
+    call: Option<Call>,
+    /// The hash of the target dispatchable.
+    call_hash: CallHash,
+
+    /// The set of accounts, that already approved the call (via this context object), including the
+    /// author.
+    ///
+    /// `approvers.len() < party.threshold` always holds.
+    approvers: HashSet<AccountId>,
+}
+
+impl Context {
```

Comment:
no new method?

Reply:
I added a mod-private constructor, so that it can be created only by initiating signature aggregation